fdbclient/NativeAPI.actor.cpp

```diff
@@ -1337,11 +1337,11 @@ void setNetworkOption(FDBNetworkOptions::Option option, Optional<StringRef> valu
 			std::string versionString = value.get().toString();
 
 			size_t index = 0;
-			size_t nextIndex = 0;
+			size_t nextIndex = versionString.find(';', index);
```

Comment:
I think now if there's no semicolon found it won't add anything to supportedVersions

Reply:
Yeah, previous behavior was UB if versionString format wasn't `(...;)+`.

Reply:
Was this undefined behavior? `find` will return `std::string::npos` if no semicolon is found, which appears to be defined as -1 (i.e. max value because it is an unsigned type). Passing `nextIndex-index` as the length to `substr` when `nextIndex==-1` will give you a value at least as large as the remainder of the string after `index` based on the maximum length of a string. `substr` is ok to take lengths that extend pass the end of the string, in which case it stops at the end of the string.

Is my above analysis correct?

Reply:
Interesting. It appears to be an invariant that versionString format is `(...;)+`, so maybe we should ASSERT that instead of handling the no semicolon case (thanks for the patience, I'm not super familiar with this code and I'm not sure who the original author is)

FWIW I'm not sure how there is UB in the old implementation if there's no semicolon. nextIndex gets assigned to [the largest value representable as size_t](http://www.cplusplus.com/reference/string/string/npos/), and [`versionString.substr(0, nextIndex-0)` will stop at the end of the string if the length is larger than nextIndex](http://www.cplusplus.com/reference/string/string/substr/), and then `index = nextIndex + 1;` is well defined to be 0 since these are unsigned types, and then we exit the loop

Reply:
> It appears to be an invariant that versionString format is `(...;)+`, so maybe we should ASSERT that instead of handling the no semicolon case

I don't think it is the case that there will be a trailing semicolon in this string, and it should be possible that the string will contain only one entry and thus no semicolon at all.

Reply:
It seems we always add a trailing semicolon here: https://github.com/apple/foundationdb/blob/7bcabbe4be9d44b020e828ea2ffd20d2ba336fce/fdbclient/MultiVersionTransaction.actor.cpp#L1276

Reply:
The block right below that will either add the local version to the end of the string without a semicolon, or it will strip off the last semicolon.

Reply:
I didn't know substr was protected against overflow. Previous code was correct then.